File: backend/app/game/manager.py
from __future__ import annotations

import asyncio
import dataclasses
import json
import uuid

from app.core.redis_bus import RedisBus
from app.game.room import Room
from app.game.typing_engine import (
    DEFAULT_TEXT_LINE_COUNT,
    build_race_paragraph,
    normalize_text_line_count,
)


@dataclasses.dataclass
class _LogicalRoomConfig:
    relay_mode: bool
    host_player_id: str | None = None
    started: bool = False
    paused: bool = False
    finished: bool = False
    race_round: int = 0
    restart_in_progress: bool = False
    text_line_count: int = DEFAULT_TEXT_LINE_COUNT


class RoomManager:
    """In-memory rooms: one *logical* URL room_id maps to many shard instances."""

    def __init__(
        self,
        *,
        redis_bus: RedisBus,
        broadcast_interval_s: float,
        max_players_per_shard: int,
        instance_id: str,
        idle_seconds: float = 45.0,
        server_ping_interval_s: float = 10.0,
    ) -> None:
        self.redis_bus = redis_bus
        self.broadcast_interval_s = broadcast_interval_s
        self.max_players_per_shard = max_players_per_shard
        self.instance_id = instance_id
        self.idle_seconds = idle_seconds
        self.server_ping_interval_s = server_ping_interval_s
        self._rooms_lock = asyncio.Lock()
        self._rooms: dict[str, Room] = {}
        self._rooms_by_logical: dict[str, list[str]] = {}
        self._logical_lock = asyncio.Lock()
        self._logical_config: dict[str, _LogicalRoomConfig] = {}

    def new_player_id(self) -> str:
        return str(uuid.uuid4())

    async def relay_mode_for(self, logical_room_id: str, relay: bool | None) -> bool:
        """First join to a logical room locks relay mode for all shards of that room."""
        async with self._logical_lock:
            c = self._logical_config.get(logical_room_id)
            if c is None:
                rm = bool(relay) if relay is not None else False
                self._logical_config[logical_room_id] = _LogicalRoomConfig(
                    relay_mode=rm,
                    text_line_count=DEFAULT_TEXT_LINE_COUNT,
                )
                return rm
            return c.relay_mode

    async def player_room_flags(
        self, logical_room_id: str, player_id: str, wants_host: bool | None
    ) -> tuple[bool, bool, bool, bool]:
        """Return (is_host, race_started, paused, finished)."""
        async with self._logical_lock:
            c = self._logical_config.get(logical_room_id)
            if c is None:
                return False, False, False, False
            if wants_host and c.host_player_id is None:
                c.host_player_id = player_id
            is_host = c.host_player_id == player_id
            return is_host, c.started, c.paused, c.finished

    async def start_logical_race(self, logical_room_id: str, player_id: str) -> bool:
        async with self._logical_lock:
            c = self._logical_config.get(logical_room_id)
            if c is None or c.started or c.host_player_id != player_id:
                return False
            c.started = True
            c.paused = False
        rooms: list[Room] = []
        async with self._rooms_lock:
            for sk in self._rooms_by_logical.get(logical_room_id, []):
                room = self._rooms.get(sk)
                if room is not None:
                    rooms.append(room)
        for room in rooms:
            await room.notify_race_started()
        return True

    async def typing_blocked(self, logical_room_id: str) -> bool:
        """True when keys / relay_pass should be ignored (not started, paused, or finished)."""
        async with self._logical_lock:
            c = self._logical_config.get(logical_room_id)
            if c is None:
                return True
            if not c.started or c.finished or c.paused:
                return True
            return False

    async def logical_play_flags(self, logical_room_id: str) -> tuple[bool, bool]:
        """Return (paused, finished) for the logical room."""
        async with self._logical_lock:
            c = self._logical_config.get(logical_room_id)
            if c is None:
                return False, False
            return c.paused, c.finished

    async def pause_logical_game(self, logical_room_id: str, player_id: str) -> bool:
        async with self._logical_lock:
            c = self._logical_config.get(logical_room_id)
            if c is None or c.host_player_id != player_id or c.finished:
                return False
            if c.paused:
                return True
            c.paused = True
        return True

    async def resume_logical_game(self, logical_room_id: str, player_id: str) -> bool:
        async with self._logical_lock:
            c = self._logical_config.get(logical_room_id)
            if c is None or c.host_player_id != player_id or c.finished:
                return False
            if not c.paused:
                return True
            c.paused = False
        return True

    async def finish_logical_game(self, logical_room_id: str, player_id: str) -> bool:
        """Idempotent: first successful call wins; duplicates return False."""
        async with self._logical_lock:
            c = self._logical_config.get(logical_room_id)
            if c is None or c.host_player_id != player_id or not c.started:
                return False
            if c.finished:
                return False
            if c.restart_in_progress:
                return False
            c.finished = True
            c.paused = False
        return True

    async def begin_race_restart(self, logical_room_id: str, player_id: str) -> bool:
        """Same-room rematch: reset shards, broadcast restart + countdown, then race_started."""
        async with self._logical_lock:
            c = self._logical_config.get(logical_room_id)
            if c is None or c.host_player_id != player_id:
                return False
            if c.restart_in_progress:
                return False
            c.restart_in_progress = True
            c.finished = False
            c.paused = False
            c.started = False
            c.race_round += 1
            c.text_line_count = normalize_text_line_count(c.text_line_count)
            rr = c.race_round
            lc = c.text_line_count
        text = build_race_paragraph(logical_room_id, race_round=rr, line_count=lc)
        rooms = await self.rooms_for_logical(logical_room_id)
        if not rooms:
            async with self._logical_lock:
                if c2 := self._logical_config.get(logical_room_id):
                    c2.restart_in_progress = False
            return False
        try:
            for r in rooms:
                await r.reset_race_session(text)
        except Exception:
            async with self._logical_lock:
                if c3 := self._logical_config.get(logical_room_id):
                    c3.restart_in_progress = False
            return False
        await self.broadcast_to_logical(
            logical_room_id,
            json.dumps({"type": "game_restart", "payload": {"text": text}}),
        )
        asyncio.create_task(
            self._restart_countdown_then_go(logical_room_id),
            name=f"restart-{logical_room_id}",
        )
        return True

    async def _restart_countdown_then_go(self, logical_room_id: str) -> None:
        try:
            for n in (3, 2, 1):
                await self.broadcast_to_logical(
                    logical_room_id,
                    json.dumps({"type": "game_restart_countdown", "payload": {"n": n}}),
                )
                await asyncio.sleep(1.0)
            await self.broadcast_to_logical(
                logical_room_id,
                json.dumps({"type": "game_restart_countdown", "payload": {"go": True}}),
            )
            await asyncio.sleep(0.45)
            async with self._logical_lock:
                c = self._logical_config.get(logical_room_id)
                if c is not None:
                    c.started = True
            rooms = await self.rooms_for_logical(logical_room_id)
            for room in rooms:
                await room.notify_race_started()
        finally:
            async with self._logical_lock:
                c = self._logical_config.get(logical_room_id)
                if c is not None:
                    c.restart_in_progress = False

    async def transfer_host_if_leaving(self, logical_room_id: str, leaving_player_id: str) -> None:
        """If the leaving player was host, assign another connected player or clear host."""
        new_host: str | None = None
        async with self._logical_lock:
            c = self._logical_config.get(logical_room_id)
            if c is None or c.host_player_id != leaving_player_id:
                return
            c.host_player_id = None

        async with self._rooms_lock:
            for sk in self._rooms_by_logical.get(logical_room_id, []):
                room = self._rooms.get(sk)
                if room is None:
                    continue
                ids = await room.list_player_ids()
                for pid in ids:
                    if pid != leaving_player_id:
                        new_host = pid
                        break
                if new_host is not None:
                    break

        async with self._logical_lock:
            c = self._logical_config.get(logical_room_id)
            if c is not None and new_host is not None:
                c.host_player_id = new_host

        await self.broadcast_host_state(logical_room_id)

    async def broadcast_to_logical(self, logical_room_id: str, message: str) -> None:
        rooms = await self.rooms_for_logical(logical_room_id)
        for room in rooms:
            await room.send_raw_to_all(message)

    async def broadcast_host_state(self, logical_room_id: str) -> None:
        async with self._logical_lock:
            c = self._logical_config.get(logical_room_id)
            host_id = c.host_player_id if c else None
        msg = json.dumps(
            {"type": "game_state_update", "payload": {"host_player_id": host_id}}
        )
        await self.broadcast_to_logical(logical_room_id, msg)

    async def logical_text_line_count(self, logical_room_id: str) -> int:
        async with self._logical_lock:
            c = self._logical_config.get(logical_room_id)
            if c is None:
                return DEFAULT_TEXT_LINE_COUNT
            n = normalize_text_line_count(c.text_line_count)
            if c.text_line_count != n:
                c.text_line_count = n
            return n

    async def canonical_paragraph_for_shard(self, room: Room) -> str:
        """Paragraph for this shard: matches siblings mid-race, or built from lobby settings."""
        logical_room_id = room.logical_room_id
        async with self._logical_lock:
            c = self._logical_config.get(logical_room_id)
            started = bool(c and c.started)
            finished = bool(c and c.finished)
            rr = c.race_round if c else 0
            lc = (
                normalize_text_line_count(c.text_line_count)
                if c
                else DEFAULT_TEXT_LINE_COUNT
            )
        if started and not finished:
            rooms = await self.rooms_for_logical(logical_room_id)
            for r in rooms:
                if r is not room and len(r.paragraph) > 0:
                    return r.paragraph
        return build_race_paragraph(logical_room_id, race_round=rr, line_count=lc)

    async def resync_room_paragraph_after_join(self, room: Room) -> None:
        text = await self.canonical_paragraph_for_shard(room)
        await room.set_shared_paragraph(text)

    async def apply_host_text_line_count(
        self, logical_room_id: str, player_id: str, raw_line_count: object
    ) -> tuple[bool, int, str]:
        """Lobby-only host setting. Returns (ok, normalized_count, new_paragraph or \"\" if unchanged)."""
        try:
            raw_int = int(raw_line_count)  # type: ignore[arg-type]
        except (TypeError, ValueError):
            raw_int = DEFAULT_TEXT_LINE_COUNT
        async with self._logical_lock:
            c = self._logical_config.get(logical_room_id)
            if c is None:
                return False, DEFAULT_TEXT_LINE_COUNT, ""
            if c.host_player_id != player_id:
                return False, c.text_line_count, ""
            if c.restart_in_progress:
                return False, c.text_line_count, ""
            if c.started and not c.finished:
                return False, c.text_line_count, ""
            c.text_line_count = normalize_text_line_count(c.text_line_count)
            n = normalize_text_line_count(raw_int)
            if n == c.text_line_count:
                return True, n, ""
            c.text_line_count = n
            rr = c.race_round
        text = build_race_paragraph(logical_room_id, race_round=rr, line_count=n)
        return True, n, text

    async def push_paragraph_to_all_shards(
        self, logical_room_id: str, text: str
    ) -> None:
        for r in await self.rooms_for_logical(logical_room_id):
            await r.set_shared_paragraph(text)

    async def rooms_for_logical(self, logical_room_id: str) -> list[Room]:
        async with self._rooms_lock:
            out: list[Room] = []
            for sk in self._rooms_by_logical.get(logical_room_id, []):
                r = self._rooms.get(sk)
                if r is not None:
                    out.append(r)
            return out

    async def assign_room(self, logical_room_id: str) -> Room:
        """Pick a shard with capacity or create a new shard for this logical room."""
        async with self._rooms_lock:
            keys = list(self._rooms_by_logical.get(logical_room_id, []))
            for sk in keys:
                room = self._rooms.get(sk)
                if room is None:
                    continue
                if await room.occupancy() < self.max_players_per_shard:
                    return room
            shard_idx = len(keys)
            storage_key = f"{logical_room_id}#{shard_idx}"
            room = Room(
                storage_key=storage_key,
                logical_room_id=logical_room_id,
                shard_index=shard_idx,
                manager=self,
            )
            self._rooms[storage_key] = room
            self._rooms_by_logical.setdefault(logical_room_id, []).append(storage_key)
            return room

    async def maybe_drop_room(self, storage_key: str) -> None:
        logical_id: str | None = None
        drop_logical_config = False
        async with self._rooms_lock:
            room = self._rooms.get(storage_key)
            if room is None:
                return
            if not await room.is_empty():
                return
            logical_id = room.logical_room_id
            self._rooms.pop(storage_key, None)
            lst = self._rooms_by_logical.get(logical_id)
            if lst and storage_key in lst:
                lst.remove(storage_key)
            if lst is not None and len(lst) == 0:
                self._rooms_by_logical.pop(logical_id, None)
                drop_logical_config = True
        if drop_logical_config and logical_id is not None:
            async with self._logical_lock:
                self._logical_config.pop(logical_id, None)

    def routing_snapshot(self) -> dict:
        """For load balancers / orchestrators: shard occupancy on this instance."""
        shards: list[dict] = []
        for sk, room in self._rooms.items():
            shards.append(
                {
                    "storage_key": sk,
                    "logical_room_id": room.logical_room_id,
                    "shard_index": room.shard_index,
                    "players": room.occupancy_sync(),
                }
            )
        return {
            "instance_id": self.instance_id,
            "max_players_per_shard": self.max_players_per_shard,
            "broadcast_interval_s": self.broadcast_interval_s,
            "broadcast_hz": 1.0 / self.broadcast_interval_s,
            "shards": shards,
        }

File: server/game/manager.ts
import { v4 as uuidv4 } from "uuid";
import { GameRoom } from "./room";
import {
  buildRaceParagraph,
  DEFAULT_TEXT_LINE_COUNT,
  normalizeTextLineCount,
} from "./typing-engine";

type LogicalRoomConfig = {
  relayMode: boolean;
  hostPlayerId: string | null;
  started: boolean;
  paused: boolean;
  finished: boolean;
  raceRound: number;
  restartInProgress: boolean;
  textLineCount: number;
};

export class RoomManager {
  readonly broadcastIntervalMs: number;
  readonly maxPlayersPerShard: number;
  readonly instanceId: string;
  readonly idleSeconds: number;
  readonly serverPingIntervalS: number;

  private rooms = new Map<string, GameRoom>();
  private roomsByLogical = new Map<string, string[]>();
  private logicalConfig = new Map<string, LogicalRoomConfig>();

  constructor(opts?: {
    broadcastHz?: number;
    maxPlayersPerShard?: number;
    instanceId?: string;
    idleSeconds?: number;
    serverPingIntervalS?: number;
  }) {
    const hz = Math.max(5, Math.min(10, opts?.broadcastHz ?? 8));
    this.broadcastIntervalMs = 1000 / hz;
    this.maxPlayersPerShard = opts?.maxPlayersPerShard ?? 150;
    this.instanceId = opts?.instanceId ?? "local-1";
    this.idleSeconds = opts?.idleSeconds ?? 45;
    this.serverPingIntervalS = opts?.serverPingIntervalS ?? 10;
  }

  newPlayerId(): string {
    return uuidv4();
  }

  relayModeFor(logicalRoomId: string, relay?: boolean | null): boolean {
    let c = this.logicalConfig.get(logicalRoomId);
    if (!c) {
      const rm = relay != null ? Boolean(relay) : false;
      c = {
        relayMode: rm,
        hostPlayerId: null,
        started: false,
        paused: false,
        finished: false,
        raceRound: 0,
        restartInProgress: false,
        textLineCount: DEFAULT_TEXT_LINE_COUNT,
      };
      this.logicalConfig.set(logicalRoomId, c);
      return rm;
    }
    return c.relayMode;
  }

  playerRoomFlags(
    logicalRoomId: string,
    playerId: string,
    wantsHost?: boolean | null
  ): { isHost: boolean; started: boolean; paused: boolean; finished: boolean } {
    const c = this.logicalConfig.get(logicalRoomId);
    if (!c) return { isHost: false, started: false, paused: false, finished: false };
    if (wantsHost && !c.hostPlayerId) c.hostPlayerId = playerId;
    return {
      isHost: c.hostPlayerId === playerId,
      started: c.started,
      paused: c.paused,
      finished: c.finished,
    };
  }

  startLogicalRace(logicalRoomId: string, playerId: string): boolean {
    const c = this.logicalConfig.get(logicalRoomId);
    if (!c || c.hostPlayerId !== playerId) return false;
    if (c.started && !c.finished) return false;
    if (c.restartInProgress) return false;
    c.started = true;
    c.paused = false;
    c.finished = false;
    for (const room of this.roomsForLogical(logicalRoomId)) {
      void room.notifyRaceStarted();
    }
    return true;
  }

  typingBlocked(logicalRoomId: string): boolean {
    const c = this.logicalConfig.get(logicalRoomId);
    if (!c) return true;
    return !c.started || c.finished || c.paused;
  }

  logicalPlayFlags(logicalRoomId: string): { paused: boolean; finished: boolean } {
    const c = this.logicalConfig.get(logicalRoomId);
    if (!c) return { paused: false, finished: false };
    return { paused: c.paused, finished: c.finished };
  }

  pauseLogicalGame(logicalRoomId: string, playerId: string): boolean {
    const c = this.logicalConfig.get(logicalRoomId);
    if (!c || c.hostPlayerId !== playerId || c.finished) return false;
    c.paused = true;
    return true;
  }

  resumeLogicalGame(logicalRoomId: string, playerId: string): boolean {
    const c = this.logicalConfig.get(logicalRoomId);
    if (!c || c.hostPlayerId !== playerId || c.finished) return false;
    c.paused = false;
    return true;
  }

  finishLogicalGame(logicalRoomId: string, playerId: string): boolean {
    const c = this.logicalConfig.get(logicalRoomId);
    if (!c || c.hostPlayerId !== playerId || !c.started || c.finished || c.restartInProgress)
      return false;
    c.finished = true;
    c.paused = false;
    return true;
  }

  async beginRaceRestart(logicalRoomId: string, playerId: string): Promise<boolean> {
    const c = this.logicalConfig.get(logicalRoomId);
    if (!c || c.hostPlayerId !== playerId || c.restartInProgress) return false;
    c.restartInProgress = true;
    c.finished = false;
    c.paused = false;
    c.started = false;
    c.raceRound += 1;
    c.textLineCount = normalizeTextLineCount(c.textLineCount);
    const text = buildRaceParagraph(logicalRoomId, {
      raceRound: c.raceRound,
      lineCount: c.textLineCount,
    });
    const rooms = this.roomsForLogical(logicalRoomId);
    if (!rooms.length) {
      c.restartInProgress = false;
      return false;
    }
    for (const r of rooms) await r.resetRaceSession(text);
    this.broadcastToLogical(
      logicalRoomId,
      JSON.stringify({ type: "game_restart", payload: { text } })
    );
    void this.restartCountdownThenGo(logicalRoomId);
    return true;
  }

  private async restartCountdownThenGo(logicalRoomId: string): Promise<void> {
    try {
      for (const n of [3, 2, 1]) {
        this.broadcastToLogical(
          logicalRoomId,
          JSON.stringify({ type: "game_restart_countdown", payload: { n } })
        );
        await new Promise((r) => setTimeout(r, 1000));
      }
      this.broadcastToLogical(
        logicalRoomId,
        JSON.stringify({ type: "game_restart_countdown", payload: { go: true } })
      );
      await new Promise((r) => setTimeout(r, 450));
      const c = this.logicalConfig.get(logicalRoomId);
      if (c) c.started = true;
      for (const room of this.roomsForLogical(logicalRoomId)) {
        await room.notifyRaceStarted();
      }
    } finally {
      const c = this.logicalConfig.get(logicalRoomId);
      if (c) c.restartInProgress = false;
    }
  }

  async transferHostIfLeaving(logicalRoomId: string, leavingPlayerId: string): Promise<void> {
    const c = this.logicalConfig.get(logicalRoomId);
    if (!c || c.hostPlayerId !== leavingPlayerId) return;
    if (leavingPlayerId.startsWith("admin-host:")) return;
    c.hostPlayerId = null;
    let newHost: string | null = null;
    for (const sk of this.roomsByLogical.get(logicalRoomId) ?? []) {
      const room = this.rooms.get(sk);
      if (!room) continue;
      for (const pid of room.listPlayerIds()) {
        if (pid !== leavingPlayerId) {
          newHost = pid;
          break;
        }
      }
      if (newHost) break;
    }
    if (newHost && c) c.hostPlayerId = newHost;
    this.broadcastHostState(logicalRoomId);
  }

  broadcastToLogical(logicalRoomId: string, message: string): void {
    for (const room of this.roomsForLogical(logicalRoomId)) {
      room.sendToAll(message);
    }
  }

  broadcastHostState(logicalRoomId: string): void {
    const c = this.logicalConfig.get(logicalRoomId);
    const hostId = c?.hostPlayerId ?? null;
    this.broadcastToLogical(
      logicalRoomId,
      JSON.stringify({ type: "game_state_update", payload: { host_player_id: hostId } })
    );
  }

  logicalTextLineCount(logicalRoomId: string): number {
    const c = this.logicalConfig.get(logicalRoomId);
    if (!c) return DEFAULT_TEXT_LINE_COUNT;
    c.textLineCount = normalizeTextLineCount(c.textLineCount);
    return c.textLineCount;
  }

  async canonicalParagraphForShard(room: GameRoom): Promise<string> {
    const c = this.logicalConfig.get(room.logicalRoomId);
    const started = Boolean(c?.started);
    const finished = Boolean(c?.finished);
    const rr = c?.raceRound ?? 0;
    const lc = c ? normalizeTextLineCount(c.textLineCount) : DEFAULT_TEXT_LINE_COUNT;
    if (started && !finished) {
      for (const r of this.roomsForLogical(room.logicalRoomId)) {
        if (r !== room && r.getParagraph().length > 0) return r.getParagraph();
      }
    }
    return buildRaceParagraph(room.logicalRoomId, { raceRound: rr, lineCount: lc });
  }

  async resyncRoomParagraphAfterJoin(room: GameRoom): Promise<void> {
    const text = await this.canonicalParagraphForShard(room);
    await room.setSharedParagraph(text);
  }

  applyHostTextLineCount(
    logicalRoomId: string,
    playerId: string,
    rawLineCount: unknown
  ): { ok: boolean; count: number; newText: string } {
    const rawInt = Number(rawLineCount);
    const c = this.logicalConfig.get(logicalRoomId);
    if (!c || c.hostPlayerId !== playerId || c.restartInProgress) {
      return { ok: false, count: c?.textLineCount ?? DEFAULT_TEXT_LINE_COUNT, newText: "" };
    }
    if (c.started && !c.finished) {
      return { ok: false, count: c.textLineCount, newText: "" };
    }
    const n = normalizeTextLineCount(Number.isFinite(rawInt) ? rawInt : DEFAULT_TEXT_LINE_COUNT);
    if (n === c.textLineCount) return { ok: true, count: n, newText: "" };
    c.textLineCount = n;
    const text = buildRaceParagraph(logicalRoomId, { raceRound: c.raceRound, lineCount: n });
    return { ok: true, count: n, newText: text };
  }

  async pushParagraphToAllShards(logicalRoomId: string, text: string): Promise<void> {
    for (const r of this.roomsForLogical(logicalRoomId)) {
      await r.setSharedParagraph(text);
    }
  }

  roomsForLogical(logicalRoomId: string): GameRoom[] {
    const out: GameRoom[] = [];
    for (const sk of this.roomsByLogical.get(logicalRoomId) ?? []) {
      const r = this.rooms.get(sk);
      if (r) out.push(r);
    }
    return out;
  }

  assignRoom(logicalRoomId: string): GameRoom {
    const keys = [...(this.roomsByLogical.get(logicalRoomId) ?? [])];
    for (const sk of keys) {
      const room = this.rooms.get(sk);
      if (room && room.occupancy() < this.maxPlayersPerShard) return room;
    }
    const shardIdx = keys.length;
    const storageKey = `${logicalRoomId}#${shardIdx}`;
    const room = new GameRoom(storageKey, logicalRoomId, shardIdx, this);
    this.rooms.set(storageKey, room);
    const list = this.roomsByLogical.get(logicalRoomId) ?? [];
    list.push(storageKey);
    this.roomsByLogical.set(logicalRoomId, list);
    return room;
  }

  async maybeDropRoom(storageKey: string): Promise<void> {
    const room = this.rooms.get(storageKey);
    if (!room || !room.isEmpty()) return;
    const logicalId = room.logicalRoomId;
    this.rooms.delete(storageKey);
    const lst = this.roomsByLogical.get(logicalId) ?? [];
    const idx = lst.indexOf(storageKey);
    if (idx >= 0) lst.splice(idx, 1);
    if (!lst.length) {
      this.roomsByLogical.delete(logicalId);
      this.logicalConfig.delete(logicalId);
    }
  }

  routingSnapshot() {
    const shards = [...this.rooms.entries()].map(([sk, room]) => ({
      storage_key: sk,
      logical_room_id: room.logicalRoomId,
      shard_index: room.shardIndex,
      players: room.occupancy(),
    }));
    return {
      instance_id: this.instanceId,
      max_players_per_shard: this.maxPlayersPerShard,
      broadcast_interval_s: this.broadcastIntervalMs / 1000,
      broadcast_hz: 1000 / this.broadcastIntervalMs,
      shards,
    };
  }

  adminHostId(logicalRoomId: string): string {
    return `admin-host:${logicalRoomId}`;
  }

  /** Reserve host role for admin panel control (no WebSocket required). */
  claimAdminHost(logicalRoomId: string): void {
    let c = this.logicalConfig.get(logicalRoomId);
    if (!c) {
      c = {
        relayMode: false,
        hostPlayerId: this.adminHostId(logicalRoomId),
        started: false,
        paused: false,
        finished: false,
        raceRound: 0,
        restartInProgress: false,
        textLineCount: DEFAULT_TEXT_LINE_COUNT,
      };
      this.logicalConfig.set(logicalRoomId, c);
      return;
    }
    c.hostPlayerId = this.adminHostId(logicalRoomId);
  }

  ensureEventRoomConfig(
    logicalRoomId: string,
    opts: { textLineCount?: number; relayMode?: boolean }
  ): void {
    let c = this.logicalConfig.get(logicalRoomId);
    if (!c) {
      c = {
        relayMode: Boolean(opts.relayMode),
        hostPlayerId: this.adminHostId(logicalRoomId),
        started: false,
        paused: false,
        finished: false,
        raceRound: 0,
        restartInProgress: false,
        textLineCount: normalizeTextLineCount(opts.textLineCount ?? DEFAULT_TEXT_LINE_COUNT),
      };
      this.logicalConfig.set(logicalRoomId, c);
      return;
    }
    if (!c.started || c.finished) {
      if (opts.relayMode != null) c.relayMode = Boolean(opts.relayMode);
      if (opts.textLineCount != null) {
        c.textLineCount = normalizeTextLineCount(opts.textLineCount);
      }
    }
    if (!c.hostPlayerId) c.hostPlayerId = this.adminHostId(logicalRoomId);
  }

  getLogicalRoomLiveState(logicalRoomId: string) {
    const c = this.logicalConfig.get(logicalRoomId);
    let playerCount = 0;
    const players: { id: string; name: string; team_id: number; progress: number; wpm: number }[] = [];
    for (const room of this.roomsForLogical(logicalRoomId)) {
      playerCount += room.occupancy();
      for (const p of room.getPlayersSnapshot()) {
        players.push({
          id: p.id,
          name: p.name,
          team_id: p.teamId,
          progress: p.progress,
          wpm: p.wpm,
        });
      }
    }
    players.sort((a, b) => b.progress - a.progress || b.wpm - a.wpm);
    return {
      room_id: logicalRoomId,
      player_count: playerCount,
      started: Boolean(c?.started),
      paused: Boolean(c?.paused),
      finished: Boolean(c?.finished),
      text_line_count: c ? normalizeTextLineCount(c.textLineCount) : DEFAULT_TEXT_LINE_COUNT,
      relay_mode: Boolean(c?.relayMode),
      admin_hosting: c?.hostPlayerId === this.adminHostId(logicalRoomId),
      race_round: c?.raceRound ?? 0,
      players,
    };
  }

  adminApplyTextLineCount(logicalRoomId: string, rawLineCount: unknown): {
    ok: boolean;
    count: number;
    newText: string;
  } {
    this.claimAdminHost(logicalRoomId);
    return this.applyHostTextLineCount(
      logicalRoomId,
      this.adminHostId(logicalRoomId),
      rawLineCount
    );
  }

  async adminGameAction(
    logicalRoomId: string,
    action: "open_lobby" | "start" | "pause" | "resume" | "finish" | "restart"
  ): Promise<{ ok: boolean; detail?: string }> {
    this.claimAdminHost(logicalRoomId);
    const hostId = this.adminHostId(logicalRoomId);

    if (action === "open_lobby") {
      const c = this.logicalConfig.get(logicalRoomId);
      if (c) {
        c.started = false;
        c.paused = false;
        c.finished = false;
        c.restartInProgress = false;
        const text = buildRaceParagraph(logicalRoomId, {
          raceRound: c.raceRound,
          lineCount: c.textLineCount,
        });
        await this.pushParagraphToAllShards(logicalRoomId, text);
      }
      this.broadcastHostState(logicalRoomId);
      return { ok: true };
    }

    if (action === "start") {
      const c = this.logicalConfig.get(logicalRoomId);
      if (c?.started && !c.finished) {
        return { ok: false, detail: "already_started" };
      }
      if (c?.finished) {
        c.finished = false;
        c.started = false;
        c.paused = false;
        c.raceRound += 1;
        const text = buildRaceParagraph(logicalRoomId, {
          raceRound: c.raceRound,
          lineCount: c.textLineCount,
        });
        for (const room of this.roomsForLogical(logicalRoomId)) {
          await room.resetRaceSession(text);
        }
        this.broadcastToLogical(
          logicalRoomId,
          JSON.stringify({ type: "room_settings_sync", payload: { text } })
        );
      }
      if (!this.startLogicalRace(logicalRoomId, hostId)) {
        return { ok: false, detail: "start_rejected" };
      }
      return { ok: true };
    }

    if (action === "pause") {
      if (!this.pauseLogicalGame(logicalRoomId, hostId)) {
        return { ok: false, detail: "pause_rejected" };
      }
      this.broadcastToLogical(
        logicalRoomId,
        JSON.stringify({ type: "game_paused", payload: {} })
      );
      return { ok: true };
    }

    if (action === "resume") {
      if (!this.resumeLogicalGame(logicalRoomId, hostId)) {
        return { ok: false, detail: "resume_rejected" };
      }
      this.broadcastToLogical(
        logicalRoomId,
        JSON.stringify({ type: "game_resumed", payload: {} })
      );
      return { ok: true };
    }

    if (action === "finish") {
      if (!this.finishLogicalGame(logicalRoomId, hostId)) {
        return { ok: false, detail: "finish_rejected" };
      }
      for (const shard of this.roomsForLogical(logicalRoomId)) {
        await shard.broadcastRaceFinishedSnapshot();
      }
      return { ok: true };
    }

    if (action === "restart") {
      if (!(await this.beginRaceRestart(logicalRoomId, hostId))) {
        return { ok: false, detail: "restart_rejected" };
      }
      return { ok: true };
    }

    return { ok: false, detail: "unknown_action" };
  }
}

type GlobalWithRoomManager = typeof globalThis & {
  __typingRaceRoomManager?: RoomManager;
};

export function getRoomManager(): RoomManager {
  const g = globalThis as GlobalWithRoomManager;
  if (!g.__typingRaceRoomManager) {
    g.__typingRaceRoomManager = new RoomManager({
      broadcastHz: Number(process.env.BROADCAST_HZ ?? 8),
      maxPlayersPerShard: Number(process.env.MAX_PLAYERS_PER_SHARD ?? 150),
      instanceId: process.env.INSTANCE_ID ?? "local-1",
    });
  }
  return g.__typingRaceRoomManager;
}

export function setRoomManager(manager: RoomManager): void {
  (globalThis as GlobalWithRoomManager).__typingRaceRoomManager = manager;
}
